File: mobile/packages/native_video_editor/android/src/main/kotlin/io/ente/native_video_editor/NativeVideoEditorPlugin.kt
package io.ente.native_video_editor

import android.util.Log
import androidx.annotation.NonNull
import androidx.media3.common.util.UnstableApi
import androidx.media3.transformer.ExportException
import io.ente.native_video_editor.core.VideoExportEngine
import io.ente.native_video_editor.core.VideoFrameExtractor
import io.ente.native_video_editor.core.VideoMetadataReader
import io.ente.native_video_editor.core.VideoCrop
import io.ente.native_video_editor.core.VideoEditRequest
import io.ente.native_video_editor.core.VideoEditResult
import io.ente.native_video_editor.core.VideoFrameExtractionResult
import io.ente.native_video_editor.core.VideoFramePolicy
import io.ente.native_video_editor.core.VideoFrameRequest
import io.ente.native_video_editor.core.VideoMetadata
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.CoroutineStart
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.cancelAndJoin
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.util.UUID

@UnstableApi
class NativeVideoEditorPlugin : FlutterPlugin, MethodCallHandler, EventChannel.StreamHandler {
    private lateinit var channel: MethodChannel
    private lateinit var progressChannel: EventChannel
    private var progressEventSink: EventChannel.EventSink? = null
    private var scope = newPluginScope()
    private var currentExportJob: Job? = null
    private val frameJobs = mutableMapOf<String, Job>()
    private lateinit var metadataReader: VideoMetadataReader
    private lateinit var frameExtractor: VideoFrameExtractor
    private lateinit var videoExporter: VideoExportEngine

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        if (!scope.isActive) scope = newPluginScope()

        channel = MethodChannel(binding.binaryMessenger, METHOD_CHANNEL)
        channel.setMethodCallHandler(this)
        progressChannel = EventChannel(binding.binaryMessenger, PROGRESS_CHANNEL)
        progressChannel.setStreamHandler(this)

        metadataReader = VideoMetadataReader()
        frameExtractor = VideoFrameExtractor(binding.applicationContext, metadataReader)
        videoExporter = VideoExportEngine(binding.applicationContext, metadataReader)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        val pluginScope = scope
        if (!pluginScope.isActive) {
            result.error("PLUGIN_DETACHED", "Native video editor is detached", null)
            return
        }

        when (call.method) {
            "getVideoInfo" -> inspectVideo(call, result, pluginScope)
            "extractFrame" -> extractSingleFrame(call, result, pluginScope)
            "extractTimeline" -> extractTimeline(call, result, pluginScope)
            "cancelFrameExtraction" -> cancelFrameExtraction(call, result, pluginScope)
            "processVideo" -> processVideo(call, result, pluginScope)
            "cancelProcessing" -> cancelProcessing(result, pluginScope)
            else -> result.notImplemented()
        }
    }

    private fun inspectVideo(call: MethodCall, result: Result, pluginScope: CoroutineScope) {
        val videoPath = call.argument<String>("videoPath")
        if (videoPath.isNullOrBlank()) {
            result.error("INVALID_ARGS", "Missing video path", null)
            return
        }

        pluginScope.launch {
            try {
                val info = withContext(Dispatchers.IO) { metadataReader.read(videoPath) }
                if (canReply(pluginScope)) result.success(info.toChannelMap())
            } catch (error: LinkageError) {
                replyError(result, "INFO_LINKAGE_ERROR", "Get video info", error, pluginScope)
            } catch (error: Exception) {
                replyError(result, "INFO_ERROR", "Get video info", error, pluginScope)
            }
        }
    }

    private fun extractSingleFrame(
        call: MethodCall,
        result: Result,
        pluginScope: CoroutineScope
    ) {
        val outputPath = call.argument<String>("outputPath")
        val positionMs = call.argument<Number>("positionMs")?.toLong()
        if (outputPath.isNullOrBlank() || positionMs == null) {
            result.error("INVALID_ARGS", "Invalid frame arguments", null)
            return
        }
        startFrameExtraction(
            call,
            result,
            UUID.randomUUID().toString(),
            listOf(outputPath),
            listOf(positionMs),
            pluginScope
        )
    }

    private fun extractTimeline(
        call: MethodCall,
        result: Result,
        pluginScope: CoroutineScope
    ) {
        val requestId = call.argument<String>("requestId")
        val outputPaths = call.argument<List<String>>("outputPaths")
        val positionsMs = call.argument<List<Number>>("positionsMs")?.map(Number::toLong)
        if (requestId.isNullOrBlank() || outputPaths == null || positionsMs == null) {
            result.error("INVALID_ARGS", "Invalid timeline arguments", null)
            return
        }
        startFrameExtraction(call, result, requestId, outputPaths, positionsMs, pluginScope)
    }

    private fun startFrameExtraction(
        call: MethodCall,
        result: Result,
        requestId: String,
        outputPaths: List<String>,
        positionsMs: List<Long>,
        pluginScope: CoroutineScope
    ) {
        if (frameJobs.containsKey(requestId)) {
            result.error("DUPLICATE_REQUEST", "Frame request ID is already active", requestId)
            return
        }
        val request = parseFrameRequest(call, outputPaths, positionsMs)
        if (request == null) {
            result.error("INVALID_ARGS", "Invalid frame extraction arguments", null)
            return
        }

        val job = pluginScope.launch(start = CoroutineStart.LAZY) {
            try {
                val extraction = frameExtractor.extract(request)
                if (canReply(pluginScope)) result.success(extraction.toChannelMap())
            } catch (error: CancellationException) {
                replyError(
                    result,
                    "FRAME_CANCELLED",
                    "Frame extraction was cancelled",
                    error,
                    pluginScope,
                    requestId
                )
            } catch (error: LinkageError) {
                replyError(
                    result,
                    "FRAME_LINKAGE_ERROR",
                    "Frame extraction",
                    error,
                    pluginScope
                )
            } catch (error: Exception) {
                replyError(
                    result,
                    "FRAME_EXTRACTION_ERROR",
                    "Frame extraction",
                    error,
                    pluginScope
                )
            } finally {
                frameJobs.remove(requestId)
            }
        }
        frameJobs[requestId] = job
        job.start()
    }

    private fun cancelFrameExtraction(
        call: MethodCall,
        result: Result,
        pluginScope: CoroutineScope
    ) {
        val requestId = call.argument<String>("requestId")
        if (requestId.isNullOrBlank()) {
            result.error("INVALID_ARGS", "Missing frame request ID", null)
            return
        }
        val job = frameJobs[requestId]
        if (job == null) {
            result.success(null)
            return
        }
        pluginScope.launch {
            job.cancelAndJoin()
            if (canReply(pluginScope)) result.success(null)
        }
    }

    private fun processVideo(call: MethodCall, result: Result, pluginScope: CoroutineScope) {
        if (currentExportJob?.isActive == true) {
            result.error("INVALID_ARGS", "Invalid or concurrent export request", null)
            return
        }
        val request = parseEditRequest(call)
        if (request == null) {
            result.error("INVALID_ARGS", "Invalid video processing arguments", null)
            return
        }

        val job = pluginScope.launch(start = CoroutineStart.LAZY) {
            try {
                val export = videoExporter.processVideo(request) { progress ->
                    if (canReply(pluginScope)) progressEventSink?.success(progress)
                }
                if (canReply(pluginScope)) result.success(export.toChannelMap())
            } catch (error: LinkageError) {
                replyError(
                    result,
                    "PROCESS_LINKAGE_ERROR",
                    "Process video",
                    error,
                    pluginScope
                )
            } catch (error: CancellationException) {
                replyError(
                    result,
                    "PROCESS_CANCELLED",
                    "Process video",
                    error,
                    pluginScope
                )
            } catch (error: Exception) {
                replyError(result, "PROCESS_ERROR", "Process video", error, pluginScope)
            } finally {
                if (currentExportJob == coroutineContext[Job]) currentExportJob = null
            }
        }
        currentExportJob = job
        job.start()
    }

    private fun cancelProcessing(result: Result, pluginScope: CoroutineScope) {
        val job = currentExportJob
        if (job == null) {
            result.success(null)
            return
        }
        pluginScope.launch {
            job.cancelAndJoin()
            if (canReply(pluginScope)) result.success(null)
        }
    }

    private fun parseFrameRequest(
        call: MethodCall,
        outputPaths: List<String>,
        positionsMs: List<Long>
    ): VideoFrameRequest? = runCatching {
        VideoFrameRequest(
            inputPath = requireNotNull(call.argument<String>("inputPath")),
            outputPaths = outputPaths,
            positionsMs = positionsMs,
            maxWidth = requireNotNull(call.argument<Int>("maxWidth")),
            maxHeight = requireNotNull(call.argument<Int>("maxHeight")),
            quality = requireNotNull(call.argument<Int>("quality")),
            policy = when (call.argument<String>("policy")) {
                "precise" -> VideoFramePolicy.PRECISE
                "nearestSync" -> VideoFramePolicy.NEAREST_SYNC
                else -> error("Invalid frame policy")
            }
        )
    }.getOrNull()

    private fun parseEditRequest(call: MethodCall): VideoEditRequest? = runCatching {
        val cropValues = listOf(
            call.argument<Int>("cropX"),
            call.argument<Int>("cropY"),
            call.argument<Int>("cropWidth"),
            call.argument<Int>("cropHeight")
        )
        require(cropValues.all { it == null } || cropValues.all { it != null })
        val crop = if (cropValues.first() == null) {
            null
        } else {
            VideoCrop(
                cropValues[0]!!,
                cropValues[1]!!,
                cropValues[2]!!,
                cropValues[3]!!
            )
        }
        VideoEditRequest(
            inputPath = requireNotNull(call.argument<String>("inputPath")),
            outputPath = requireNotNull(call.argument<String>("outputPath")),
            trimStartMs = call.argument<Number>("trimStartMs")?.toLong(),
            trimEndMs = call.argument<Number>("trimEndMs")?.toLong(),
            rotateDegrees = call.argument<Int>("rotateDegrees"),
            crop = crop
        )
    }.getOrNull()

    private suspend fun replyError(
        result: Result,
        code: String,
        operation: String,
        error: Throwable,
        pluginScope: CoroutineScope,
        details: Any? = null
    ) {
        Log.e(TAG, "$operation failed", error)
        withContext(NonCancellable + Dispatchers.Main.immediate) {
            if (canReply(pluginScope)) {
                result.error(
                    code,
                    if (code.endsWith("CANCELLED")) operation else error.message ?: operation,
                    details ?: buildErrorDetails(error)
                )
            }
        }
    }

    private fun canReply(pluginScope: CoroutineScope): Boolean = pluginScope.isActive

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        progressEventSink = null
        channel.setMethodCallHandler(null)
        progressChannel.setStreamHandler(null)
        scope.cancel()
        currentExportJob = null
        frameJobs.clear()
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        if (scope.isActive) progressEventSink = events
    }

    override fun onCancel(arguments: Any?) {
        progressEventSink = null
    }

    private fun VideoMetadata.toChannelMap(): Map<String, Any> = buildMap {
        put("duration", durationMs)
        put("width", width)
        put("height", height)
        put("displayWidth", displayWidth)
        put("displayHeight", displayHeight)
        put("rotation", rotationDegrees)
        bitrate?.let { put("bitrate", it) }
        frameRate?.let { put("frameRate", it) }
    }

    private fun VideoFrameExtractionResult.toChannelMap(): Map<String, Any> = mapOf(
        "videoInfo" to videoInfo.toChannelMap(),
        "frames" to frames.map {
            mapOf("outputPath" to it.outputPath, "width" to it.width, "height" to it.height)
        }
    )

    private fun VideoEditResult.toChannelMap(): Map<String, Any> = mapOf(
        "outputPath" to outputPath,
        "isReEncoded" to isReEncoded
    )

    private fun buildErrorDetails(error: Throwable): Map<String, Any?> {
        val details = mutableMapOf<String, Any?>(
            "type" to error::class.java.simpleName,
            "message" to (error.message ?: "")
        )
        val exportException = when {
            error is ExportException -> error
            error.cause is ExportException -> error.cause as ExportException
            else -> null
        }
        exportException?.let { details["errorCode"] = it.errorCode }
        (error.cause ?: exportException?.cause)?.let {
            details["causeType"] = it::class.java.simpleName
            details["causeMessage"] = it.message ?: ""
        }
        return details
    }

    private companion object {
        const val TAG = "NativeVideoEditorPlugin"
        const val METHOD_CHANNEL = "native_video_editor"
        const val PROGRESS_CHANNEL = "native_video_editor/progress"

        fun newPluginScope() = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    }
}

File: mobile/packages/native_video_editor/ios/Classes/NativeVideoEditorPlugin.swift
@preconcurrency import Flutter
import Foundation

@MainActor
public final class NativeVideoEditorPlugin: NSObject, @preconcurrency FlutterPlugin {
    private var isDetached = false
    private var progressEventSink: FlutterEventSink?
    private var frameTasks: [String: Task<Void, Never>] = [:]
    private var frameCancellationResults: [String: FlutterResult] = [:]
    private var exportTask: Task<Void, Never>?
    private var exportCancellationResult: FlutterResult?
    private let metadataReader = VideoMetadataReader()
    private lazy var frameExtractor = VideoFrameExtractor(metadataReader: metadataReader)
    private let videoExporter = VideoExportEngine()

    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(
            name: "native_video_editor",
            binaryMessenger: registrar.messenger()
        )
        let instance = NativeVideoEditorPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
        registrar.publish(instance)

        let progressChannel = FlutterEventChannel(
            name: "native_video_editor/progress",
            binaryMessenger: registrar.messenger()
        )
        progressChannel.setStreamHandler(instance)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard !isDetached else {
            result(flutterError("PLUGIN_DETACHED", message: "Native video editor is detached"))
            return
        }
        switch call.method {
        case "getVideoInfo":
            inspectVideo(call, result: result)
        case "extractFrame":
            extractSingleFrame(call, result: result)
        case "extractTimeline":
            extractTimeline(call, result: result)
        case "cancelFrameExtraction":
            cancelFrameExtraction(call, result: result)
        case "processVideo":
            processVideo(call, result: result)
        case "cancelProcessing":
            cancelProcessing(result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func inspectVideo(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let args = call.arguments as? [String: Any],
            let videoPath = args["videoPath"] as? String,
            !videoPath.isEmpty
        else {
            result(flutterError("INVALID_ARGS", message: "Invalid arguments"))
            return
        }

        Task { [weak self] in
            guard let self else { return }
            do {
                let info = try await metadataReader.read(
                    from: URL(fileURLWithPath: videoPath)
                )
                guard !isDetached else { return }
                result(info.channelValue)
            } catch {
                guard !isDetached else { return }
                result(flutterError(error, defaultCode: "INFO_ERROR"))
            }
        }
    }

    private func extractSingleFrame(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        guard let args = call.arguments as? [String: Any],
            let outputPath = args["outputPath"] as? String,
            !outputPath.isEmpty,
            let positionMs = int64(args["positionMs"])
        else {
            result(flutterError("INVALID_ARGS", message: "Invalid frame arguments"))
            return
        }
        startFrameExtraction(
            args: args,
            requestID: UUID().uuidString,
            outputPaths: [outputPath],
            positionsMs: [positionMs],
            result: result
        )
    }

    private func extractTimeline(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        guard let args = call.arguments as? [String: Any],
            let requestID = args["requestId"] as? String,
            !requestID.isEmpty,
            let outputPaths = args["outputPaths"] as? [String],
            outputPaths.allSatisfy({ !$0.isEmpty }),
            let positions = args["positionsMs"] as? [NSNumber]
        else {
            result(flutterError("INVALID_ARGS", message: "Invalid timeline arguments"))
            return
        }
        startFrameExtraction(
            args: args,
            requestID: requestID,
            outputPaths: outputPaths,
            positionsMs: positions.map(\.int64Value),
            result: result
        )
    }

    private func startFrameExtraction(
        args: [String: Any],
        requestID: String,
        outputPaths: [String],
        positionsMs: [Int64],
        result: @escaping FlutterResult
    ) {
        guard frameTasks[requestID] == nil else {
            result(
                flutterError(
                    "DUPLICATE_REQUEST",
                    message: "Frame request ID is already active",
                    details: requestID
                ))
            return
        }
        let request: VideoFrameRequest
        do {
            request = try makeFrameRequest(
                args: args,
                outputPaths: outputPaths,
                positionsMs: positionsMs
            )
        } catch {
            result(flutterError(error, defaultCode: "INVALID_ARGS"))
            return
        }

        let task = Task { [weak self] in
            guard let self else { return }
            let value: Any
            do {
                let extraction = try await frameExtractor.extract(request)
                value = extraction.channelValue(outputPaths: outputPaths)
            } catch is CancellationError {
                value = flutterError(
                    "FRAME_CANCELLED",
                    message: "Frame extraction was cancelled",
                    details: requestID
                )
            } catch {
                value = flutterError(error, defaultCode: "FRAME_EXTRACTION_ERROR")
            }
            finishFrameRequest(requestID: requestID, result: result, value: value)
        }
        frameTasks[requestID] = task
    }

    private func cancelFrameExtraction(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        guard let args = call.arguments as? [String: Any],
            let requestID = args["requestId"] as? String
        else {
            result(flutterError("INVALID_ARGS", message: "Missing frame request ID"))
            return
        }
        guard let task = frameTasks[requestID] else {
            result(nil)
            return
        }
        guard frameCancellationResults[requestID] == nil else {
            result(nil)
            return
        }
        frameCancellationResults[requestID] = result
        task.cancel()
    }

    private func finishFrameRequest(
        requestID: String,
        result: @escaping FlutterResult,
        value: Any
    ) {
        guard frameTasks.removeValue(forKey: requestID) != nil else { return }
        let cancellationResult = frameCancellationResults.removeValue(forKey: requestID)
        guard !isDetached else { return }
        result(value)
        cancellationResult?(nil)
    }

    private func processVideo(
        _ call: FlutterMethodCall,
        result: @escaping FlutterResult
    ) {
        guard exportTask == nil else {
            result(flutterError("CONCURRENT_EXPORT", message: "A video export is already active"))
            return
        }
        guard let args = call.arguments as? [String: Any],
            let outputPath = args["outputPath"] as? String,
            !outputPath.isEmpty
        else {
            result(flutterError("INVALID_ARGS", message: "Invalid arguments"))
            return
        }
        let request: VideoEditRequest
        do {
            request = try makeEditRequest(args: args)
        } catch {
            result(flutterError(error, defaultCode: "INVALID_ARGS"))
            return
        }

        let task = Task { [weak self] in
            guard let self else { return }
            let value: Any
            do {
                let export = try await videoExporter.export(request) { [weak self] progress in
                    Task { @MainActor [weak self] in
                        guard let self, !isDetached else { return }
                        progressEventSink?(progress)
                    }
                }
                value = [
                    "outputPath": outputPath,
                    "isReEncoded": export.isReEncoded,
                ]
            } catch is CancellationError {
                value = flutterError("PROCESS_CANCELLED", message: "Export cancelled")
            } catch {
                value = flutterError(error, defaultCode: "PROCESS_ERROR")
            }
            finishExport(result: result, value: value)
        }
        exportTask = task
    }

    private func cancelProcessing(result: @escaping FlutterResult) {
        guard let exportTask else {
            result(nil)
            return
        }
        guard exportCancellationResult == nil else {
            result(nil)
            return
        }
        exportCancellationResult = result
        exportTask.cancel()
    }

    private func finishExport(result: @escaping FlutterResult, value: Any) {
        guard exportTask != nil else { return }
        exportTask = nil
        let cancellationResult = exportCancellationResult
        exportCancellationResult = nil
        guard !isDetached else { return }
        result(value)
        cancellationResult?(nil)
    }

    private func makeFrameRequest(
        args: [String: Any],
        outputPaths: [String],
        positionsMs: [Int64]
    ) throws -> VideoFrameRequest {
        guard let inputPath = args["inputPath"] as? String,
            !inputPath.isEmpty,
            outputPaths.allSatisfy({ !$0.isEmpty }),
            let maxWidth = int(args["maxWidth"]),
            let maxHeight = int(args["maxHeight"]),
            let quality = int(args["quality"]),
            let policyValue = args["policy"] as? String
        else {
            throw VideoEditorError.invalidRequest("Invalid frame extraction arguments")
        }
        let policy: VideoFramePolicy
        switch policyValue {
        case "precise": policy = .precise
        case "nearestSync": policy = .nearestSync
        default:
            throw VideoEditorError.invalidRequest("Invalid frame policy")
        }
        return try VideoFrameRequest(
            inputURL: URL(fileURLWithPath: inputPath),
            outputURLs: outputPaths.map { URL(fileURLWithPath: $0) },
            positionsMs: positionsMs,
            maxWidth: maxWidth,
            maxHeight: maxHeight,
            quality: quality,
            policy: policy
        )
    }

    private func makeEditRequest(args: [String: Any]) throws -> VideoEditRequest {
        guard let inputPath = args["inputPath"] as? String,
            !inputPath.isEmpty,
            let outputPath = args["outputPath"] as? String,
            !outputPath.isEmpty
        else {
            throw VideoEditorError.invalidRequest("Invalid video paths")
        }
        let cropValues = [
            int(args["cropX"]),
            int(args["cropY"]),
            int(args["cropWidth"]),
            int(args["cropHeight"]),
        ]
        guard cropValues.allSatisfy({ $0 == nil }) || cropValues.allSatisfy({ $0 != nil }) else {
            throw VideoEditorError.invalidRequest("Crop values must be provided together")
        }
        let crop =
            cropValues[0] == nil
            ? nil
            : try VideoCrop(
                x: cropValues[0]!,
                y: cropValues[1]!,
                width: cropValues[2]!,
                height: cropValues[3]!
            )
        return try VideoEditRequest(
            inputURL: URL(fileURLWithPath: inputPath),
            outputURL: URL(fileURLWithPath: outputPath),
            trimStartMs: int64(args["trimStartMs"]),
            trimEndMs: int64(args["trimEndMs"]),
            rotateDegrees: int(args["rotateDegrees"]),
            crop: crop
        )
    }

    private func int(_ value: Any?) -> Int? {
        (value as? NSNumber)?.intValue
    }

    private func int64(_ value: Any?) -> Int64? {
        (value as? NSNumber)?.int64Value
    }

    private func flutterError(
        _ error: Error,
        defaultCode: String
    ) -> FlutterError {
        guard let videoError = error as? VideoEditorError else {
            return flutterError(
                defaultCode,
                message: error.localizedDescription,
                details: String(describing: error)
            )
        }
        let code: String
        switch videoError {
        case .invalidRequest: code = "INVALID_ARGS"
        case .metadataUnavailable: code = defaultCode
        case .frameExtraction: code = "FRAME_EXTRACTION_ERROR"
        case .frameWrite: code = "FRAME_WRITE_ERROR"
        case .invalidTimeRange: code = "INVALID_TIME_RANGE"
        case .composition: code = "COMPOSITION_ERROR"
        case .export: code = defaultCode
        }
        return flutterError(
            code,
            message: videoError.message,
            details: videoError.underlyingError.map(String.init(describing:))
        )
    }

    private func flutterError(
        _ code: String,
        message: String,
        details: Any? = nil
    ) -> FlutterError {
        FlutterError(code: code, message: message, details: details)
    }

    public func detachFromEngine(for registrar: FlutterPluginRegistrar) {
        guard !isDetached else { return }
        isDetached = true
        progressEventSink = nil
        frameTasks.values.forEach { $0.cancel() }
        frameTasks.removeAll()
        frameCancellationResults.removeAll()
        exportTask?.cancel()
        exportTask = nil
        exportCancellationResult = nil
    }
}

private extension VideoMetadata {
    var channelValue: [String: Any] {
        var value: [String: Any] = [
            "duration": durationMs,
            "width": width,
            "height": height,
            "displayWidth": displayWidth,
            "displayHeight": displayHeight,
            "rotation": rotationDegrees,
        ]
        if let bitrate { value["bitrate"] = bitrate }
        if let frameRate { value["frameRate"] = frameRate }
        return value
    }
}

private extension VideoFrameExtractionResult {
    func channelValue(outputPaths: [String]) -> [String: Any] {
        [
            "videoInfo": videoInfo.channelValue,
            "frames": zip(frames, outputPaths).map { frame, outputPath in
                [
                    "outputPath": outputPath,
                    "width": frame.width,
                    "height": frame.height,
                ] as [String: Any]
            },
        ]
    }
}

extension NativeVideoEditorPlugin: @preconcurrency FlutterStreamHandler {
    public func onListen(
        withArguments arguments: Any?,
        eventSink events: @escaping FlutterEventSink
    ) -> FlutterError? {
        if !isDetached { progressEventSink = events }
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        progressEventSink = nil
        return nil
    }
}
